### ISSUE TITLE
Server templates are using the wrong port name

### ISSUE THREAD

**Issue by mhyllander** [jcmoraisjr/haproxy-ingress#788]
**Description of the problem**

When using a DNS resolver to look up pod IP addresses, the generated server template is using the wrong port name from the kubernetes service. It is using the `targetPort` name instead of the port `name`. Lookups fail if the `targetPort` is different from `name`.

**Expected behavior**

I am using an AKS cluster with coredns as DNS server. For a service, both A and SRV records are created. The SRV record has the form `_<port>._<protocol>.<service>.<namespace>.svc.cluster.local`. The port here is the `name` from the service endpoint:

```YAML
  ports:
  - name: http-admin
    port: 9990
    protocol: TCP
    targetPort: admin
```

The SRV record created in coredns for the service endpoint in this example is `_http-admin._tcp.hivejs.platform.svc.cluster.local`, using the `name` for the port.

But the server template generated by haproxy-ingress has `_admin._tcp.hivejs.platform.svc.cluster.local`, using the `targetPort` for the port, which obviously will fail because that SRV record does not exist.

**Environment information**

HAProxy Ingress version: `v0.12.2`

**Comment by jcmoraisjr** [comment]
Hi thanks for the report! I'm addressing this using the following rules, let me know if something can be improved:

- if the target port is a number: `<svcname>.<namespace>.svc.cluster.local:<port>`
- if the target port is not a number and there is a port name: `_<portname>._tcp.<svcname>.<namespace>.svc.cluster.local`
- if the target port is not a number and there is not a port name: `_<targetport>._tcp.<svcname>.<namespace>.svc.cluster.local`

**Comment by mhyllander** [comment]
Hi! I'm not sure about the third alternative. I don't think SRV records are ever created using the target port. At least that's the way I read the k8s docs: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#srv-records. And I tried to remove the port name, which resulted in an error.

I'm thinking, since this is all about when service discovery is done using haproxy DNS lookups, don't you need to distinguish between normal and headless services, too? Something like:

* if the target port is a number:
  * if the service is headless: `<svcname>.<namespace>.svc.cluster.local:<targetport>`
  * if the service is normal: `<svcname>.<namespace>.svc.cluster.local:<port>`
* if the target port is not a number (special case to signify using SRV lookups): `_<portname>._tcp.<svcname>.<namespace>.svc.cluster.local`

**Comment by jcmoraisjr** [comment]
Fixed in v0.12.4! Closing.
